File: core/genetics/inheritance.py
"""
Inheritance System for TurboShells
Handles Mendelian inheritance between parent turtles
"""

import random
from typing import Dict, List, Tuple, Union
from .gene_definitions import GeneDefinitions


class Inheritance:
    """
    Implements Mendelian inheritance patterns for genetic traits.
    Single responsibility: Handle gene inheritance logic.
    """
    
    def __init__(self, gene_definitions: GeneDefinitions = None):
        self.gene_definitions = gene_definitions or GeneDefinitions()
    
    def inherit_genetics(self, parent1_genetics: Dict, parent2_genetics: Dict) -> Dict[str, Union[Tuple, str, float]]:
        """
        Inherit genetics from two parents with random selection
        Implements basic Mendelian inheritance (50/50 chance from each parent)
        """
        child_genetics = {}
        
        for gene_name in self.gene_definitions.get_all_gene_names():
            # Get parent values or use defaults
            parent1_value = parent1_genetics.get(gene_name, 
                self.gene_definitions.get_gene_definition(gene_name).get('default'))
            parent2_value = parent2_genetics.get(gene_name, 
                self.gene_definitions.get_gene_definition(gene_name).get('default'))
            
            # Inherit from one parent (50/50 chance)
            if random.random() < 0.5:
                child_genetics[gene_name] = parent1_value
            else:
                child_genetics[gene_name] = parent2_value
        
        return child_genetics
    
    def inherit_with_dominance(self, parent1_genetics: Dict, parent2_genetics: Dict,
                              dominant_genes: Dict[str, str] = None) -> Dict[str, Union[Tuple, str, float]]:
        """
        Inherit genetics with dominant gene patterns
        dominant_genes: {'gene_name': 'parent1' or 'parent2'}
        """
        child_genetics = {}
        dominant_genes = dominant_genes or {}
        
        for gene_name in self.gene_definitions.get_all_gene_names():
            parent1_value = parent1_genetics.get(gene_name, 
                self.gene_definitions.get_gene_definition(gene_name).get('default'))
            parent2_value = parent2_genetics.get(gene_name, 
                self.gene_definitions.get_gene_definition(gene_name).get('default'))
            
            # Check for dominance
            dominance = dominant_genes.get(gene_name)
            if dominance == 'parent1':
                child_genetics[gene_name] = parent1_value
            elif dominance == 'parent2':
                child_genetics[gene_name] = parent2_value
            else:
                # Random inheritance
                child_genetics[gene_name] = parent1_value if random.random() < 0.5 else parent2_value
        
        return child_genetics
    
    def inherit_blended(self, parent1_genetics: Dict, parent2_genetics: Dict,
                       blend_genes: List[str] = None) -> Dict[str, Union[Tuple, str, float]]:
        """
        Inherit genetics with blending for continuous values
        blend_genes: List of gene names to blend (average)
        """
        child_genetics = {}
        blend_genes = blend_genes or []
        
        for gene_name in self.gene_definitions.get_all_gene_names():
            parent1_value = parent1_genetics.get(gene_name, 
                self.gene_definitions.get_gene_definition(gene_name).get('default'))
            parent2_value = parent2_genetics.get(gene_name, 
                self.gene_definitions.get_gene_definition(gene_name).get('default'))
            
            gene_def = self.gene_definitions.get_gene_definition(gene_name)
            
            if gene_name in blend_genes and gene_def['type'] == 'continuous':
                # Blend continuous values
                child_genetics[gene_name] = (parent1_value + parent2_value) / 2
            elif gene_name in blend_genes and gene_def['type'] == 'rgb':
                # Blend colors
                child_genetics[gene_name] = tuple(
                    int((parent1_value[i] + parent2_value[i]) / 2)
                    for i in range(3)
                )
            else:
                # Random inheritance
                child_genetics[gene_name] = parent1_value if random.random() < 0.5 else parent2_value
        
        return child_genetics
    
    def inherit_color_patterns(self, parent1_genetics: Dict, parent2_genetics: Dict) -> Dict[str, Union[Tuple, str, float]]:
        """
        Specialized inheritance for color patterns with mixing
        """
        child_genetics = {}
        
        for gene_name in self.gene_definitions.get_all_gene_names():
            gene_def = self.gene_definitions.get_gene_definition(gene_name)
            parent1_value = parent1_genetics.get(gene_name, gene_def.get('default'))
            parent2_value = parent2_genetics.get(gene_name, gene_def.get('default'))
            
            if gene_def['type'] == 'rgb':
                # Color mixing with random bias
                bias = random.uniform(0.3, 0.7)  # Bias towards one parent
                child_genetics[gene_name] = tuple(
                    int(parent1_value[i] * bias + parent2_value[i] * (1 - bias))
                    for i in range(3)
                )
            elif gene_def['type'] == 'discrete' and 'pattern' in gene_name:
                # Pattern inheritance with preference
                if random.random() < 0.7:  # 70% chance to inherit pattern type
                    child_genetics[gene_name] = parent1_value if random.random() < 0.5 else parent2_value
                else:
                    # Random pattern
                    child_genetics[gene_name] = random.choice(gene_def['range'])
            else:
                # Standard inheritance
                child_genetics[gene_name] = parent1_value if random.random() < 0.5 else parent2_value
        
        return child_genetics
    
    def calculate_genetic_similarity(self, genetics1: Dict, genetics2: Dict) -> float:
        """
        Calculate similarity percentage between two genetic profiles
        """
        total_genes = len(self.gene_definitions.get_all_gene_names())
        similar_genes = 0
        
        for gene_name in self.gene_definitions.get_all_gene_names():
            gene_def = self.gene_definitions.get_gene_definition(gene_name)
            value1 = genetics1.get(gene_name, gene_def.get('default'))
            value2 = genetics2.get(gene_name, gene_def.get('default'))
            
            if gene_def['type'] == 'rgb':
                # Color similarity (Euclidean distance in RGB space)
                distance = sum((value1[i] - value2[i]) ** 2 for i in range(3)) ** 0.5
                max_distance = (255 ** 2 * 3) ** 0.5  # Maximum possible distance
                similarity = 1 - (distance / max_distance)
                if similarity > 0.8:  # Consider similar if >80% similar
                    similar_genes += 1
            elif value1 == value2:
                similar_genes += 1
        
        return similar_genes / total_genes if total_genes > 0 else 0

File: turboshells-core/src/genetics/inheritance.rs
//! Inheritance system - Mendelian genetics

use std::collections::HashMap;
use rand::Rng;
use crate::types::{GeneValue, Rgb};
use super::genes::GeneDefinitions;

/// Implements Mendelian inheritance patterns
pub struct Inheritance {
    definitions: GeneDefinitions,
}

impl Inheritance {
    pub fn new(definitions: GeneDefinitions) -> Self {
        Self { definitions }
    }
    
    /// Basic Mendelian inheritance (50/50 chance from each parent)
    pub fn inherit(
        &self,
        parent1: &HashMap<String, GeneValue>,
        parent2: &HashMap<String, GeneValue>,
    ) -> HashMap<String, GeneValue> {
        let mut rng = rand::thread_rng();
        let mut child = HashMap::new();
        
        for name in self.definitions.names() {
            let p1_value = parent1.get(name);
            let p2_value = parent2.get(name);
            
            let value = match (p1_value, p2_value) {
                (Some(v1), Some(v2)) => {
                    if rng.gen_bool(0.5) { v1.clone() } else { v2.clone() }
                },
                (Some(v), None) | (None, Some(v)) => v.clone(),
                (None, None) => {
                    if let Some(def) = self.definitions.get(name) {
                        def.default.clone()
                    } else {
                        continue;
                    }
                },
            };
            
            child.insert(name.clone(), value);
        }
        
        child
    }
    
    /// Blended inheritance (average continuous values, mix colors)
    pub fn inherit_blended(
        &self,
        parent1: &HashMap<String, GeneValue>,
        parent2: &HashMap<String, GeneValue>,
    ) -> HashMap<String, GeneValue> {
        let mut rng = rand::thread_rng();
        let mut child = HashMap::new();
        
        for name in self.definitions.names() {
            let def = match self.definitions.get(name) {
                Some(d) => d,
                None => continue,
            };
            
            let p1_value = parent1.get(name);
            let p2_value = parent2.get(name);
            
            let value = match (p1_value, p2_value, def.gene_type.as_str()) {
                // Blend RGB colors
                (Some(GeneValue::Rgb(c1)), Some(GeneValue::Rgb(c2)), "rgb") => {
                    let bias = rng.gen_range(0.3..0.7);
                    GeneValue::Rgb(c1.blend(c2, bias))
                },
                // Average continuous values
                (Some(GeneValue::Continuous(f1)), Some(GeneValue::Continuous(f2)), "continuous") => {
                    GeneValue::Continuous((f1 + f2) / 2.0)
                },
                // Discrete: random from parent
                (Some(v1), Some(v2), "discrete") => {
                    if rng.gen_bool(0.5) { v1.clone() } else { v2.clone() }
                },
                // Fallback
                (Some(v), None, _) | (None, Some(v), _) => v.clone(),
                _ => def.default.clone(),
            };
            
            child.insert(name.clone(), value);
        }
        
        child
    }
    
    /// Calculate genetic similarity (0.0 to 1.0)
    pub fn calculate_similarity(
        &self,
        genetics1: &HashMap<String, GeneValue>,
        genetics2: &HashMap<String, GeneValue>,
    ) -> f32 {
        let mut similar = 0.0;
        let mut total = 0.0;
        
        for name in self.definitions.names() {
            let v1 = genetics1.get(name);
            let v2 = genetics2.get(name);
            
            let def = match self.definitions.get(name) {
                Some(d) => d,
                None => continue,
            };
            
            total += 1.0;
            
            match (v1, v2, def.gene_type.as_str()) {
                (Some(GeneValue::Rgb(c1)), Some(GeneValue::Rgb(c2)), "rgb") => {
                    // Color similarity based on Euclidean distance
                    let max_dist = (255.0_f32.powi(2) * 3.0).sqrt();
                    let dist = c1.distance(c2);
                    similar += 1.0 - (dist / max_dist);
                },
                (Some(GeneValue::Continuous(f1)), Some(GeneValue::Continuous(f2)), "continuous") => {
                    if let Some((min, max)) = def.continuous_range {
                        let range = max - min;
                        let diff = (f1 - f2).abs() / range;
                        similar += 1.0 - diff;
                    }
                },
                (Some(GeneValue::Discrete(s1)), Some(GeneValue::Discrete(s2)), "discrete") => {
                    if s1 == s2 {
                        similar += 1.0;
                    }
                },
                _ => {},
            }
        }
        
        if total > 0.0 { similar / total } else { 0.0 }
    }
}
